### PR TITLE
feat: show camera position in bird view

### PR DESCRIPTION
**Pull request by RaymondHuang210129**
I have generate a new class `CameraComponent` which draws the camera in the bird view, and found some problems during the implementation:
1. The sphere shader should be more generalized to make it more reusable: SphereComponent should use shader program that also expect normal vector to be sent although we know the position is the normal vector in this special case. This generalized shader allows us to draw more different components including cameraComponent.
2. The current implemetation of `getQuadTreeLeaves` is problematic since it copies the node of the tree instead of the take the reference of the original node in the tree. As the workaround I change this function to return the vector of node pointers so the view in the original tree is updated. After modifying this, I found there is also a bug in `updateView`. The `xDelta` and `yDelta` is not the moving distance of the mouse but the current mouse location. If we keep add the `xDelta` to horizontal and vertical radian, the view will keep rotating and rotating. I think we did not found this bugs before because we copies the node before we update it thus this logic flaw is covered. To make the pointer safe, shared_ptr should be used but it requires some modifications to quad-tree module (e.g. avoid using struct).
3. When we split a view into two views, one of the quadTree leaf becomes parent node with two leaf nodes. I expect that one of the leaf node should be the view before it is splited so that for each split we only add a new camera position instead of generating two new positions and making original one disapeared. I feel that my thought is more reasonable base on my experience using tmux to split the view. In short, the left child of the node should have the same camera position and angle with the parent node.

Lastly, I have trouble rendering the color of cameraComponent. the color of the surface is always black and I am not sure how to modify the shader program to make it correct. Some assistance is needed. (resolved)

Update: 
Smart pointers are applied. 
Revised node growing logic so that the parent's view is retained in one of the screen after splitted.
SphereComponent uses generalized lighting shader.

https://user-images.githubusercontent.com/12983673/204065231-7581505f-dca0-4c10-a5b8-33b6fed32236.mp4

### CODE REVIEW

File: src/camera.ixx [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+private:
+  mutable GLuint _vao{};
+};
+
+export class CameraComponent {
+public:
+  CameraComponent(const glm::vec3& position) {
+    model = glm::translate(model, position);
+    model = glm::scale(model, glm::vec3{0.05f, 0.05f, 0.05f});
+  }
```
Comment (sdCaramel302812) [inline src/camera.ixx:114]: The color didn't render because you forgot to set ambient, diffuse etc

File: src/camera.ixx [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        {{-1, 1, -1}, {0, 1, 0}},
+        {{-1, 1, 1}, {0, 1, 0}},
+        {{1, 1, 1}, {0, 1, 0}}, // right y plane
+
+        {{-1, -1, 1}, {0, -1, 0}},
+        {{-1, -1, -1}, {0, -1, 0}},
+        {{1, -1, -1}, {0, -1, 0}},
+        {{1, -1, 1}, {0, -1, 0}}, // left y plane
+
+        {{1, 1, -1}, {1, 0, 0}},
```
Comment (sdCaramel302812) [inline src/camera.ixx:56]: wrong position. there should have four points to form a plane

File: src/camera.ixx [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        {{1, -1, 1}, {1, 0, 0}},
+        {{1, 1, -1}, {1, 0, 0}},
+        {{1, -1, 1}, {1, 0, 0}}, // front x plane
+
+        {{-1, 1, -1}, {-1, 0, 0}},
```
Comment (sdCaramel302812) [inline src/camera.ixx:61]: wrong position

File: src/camera.ixx [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        {{-0.5, 0.5, 1}, {0, 1, 0}},
+        {{-1, 1, 3}, {0, 1, 0}},
+        {{1, 1, 3}, {0, 1, 0}}, // right y lens hood
+
+        {{-1, -1, 3}, {0, -1, 0}},
+        {{-0.5, -0.5, 1}, {0, -1, 0}},
+        {{0.5, -0.5, 1}, {0, -1, 0}},
+        {{1, -1, 3}, {0, -1, 0}}, // left y lens hood
+
+        {{1, 1, 1}, {1, 0, 0}},
```
Comment (sdCaramel302812) [inline src/camera.ixx:76]: wrong position, wrong normal

File: src/camera.ixx [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        {{1, -1, 3}, {1, 0, 0}},
+        {{1, 1, 1}, {1, 0, 0}},
+        {{1, -1, 3}, {1, 0, 0}}, // front x lens hood
+
+        {{-0.5, 0.5, 1}, {-1, 0, 0}},
```
Comment (sdCaramel302812) [inline src/camera.ixx:81]: wrong position, wrong normal

File: src/camera.ixx [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        {{1, -1, 1}, {1, 0, 0}},
+        {{1, 1, -1}, {1, 0, 0}},
+        {{1, -1, 1}, {1, 0, 0}}, // front x plane
+
+        {{-1, 1, -1}, {-1, 0, 0}},
+        {{-1, -1, 1}, {-1, 0, 0}},
+        {{-1, 1, -1}, {-1, 0, 0}},
+        {{-1, -1, 1}, {-1, 0, 0}}, // back x plane
+
+        {{0.5, 0.5, 1}, {0, 1, 0}},
```
Comment (sdCaramel302812) [inline src/camera.ixx:66]: wrong normal. the normal should be the cross product of two vectors.
https://www.symbolab.com/solver/vector-cross-product-calculator/%5Cbegin%7Bpmatrix%7D1%260%260%5Cend%7Bpmatrix%7D%5Ctimes%5Cbegin%7Bpmatrix%7D-0.5%260.5%262%5Cend%7Bpmatrix%7D?or=input

File: src/camera.ixx [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        {{-0.5, 0.5, 1}, {0, 1, 0}},
+        {{-1, 1, 3}, {0, 1, 0}},
+        {{1, 1, 3}, {0, 1, 0}}, // right y lens hood
+
+        {{-1, -1, 3}, {0, -1, 0}},
```
Comment (sdCaramel302812) [inline src/camera.ixx:71]: wrong normal

**Review by sdCaramel302812 — CHANGES_REQUESTED** [review]

File: src/camera.ixx (hunk 1 above)
Comment (RaymondHuang210129) [reply src/camera.ixx:114]: Got it.

File: src/camera.ixx (hunk 2 above)
Comment (RaymondHuang210129) [reply src/camera.ixx:56]: Got it. Seems that WallCompoent also have this bug.

File: src/camera.ixx (hunk 6 above)
Comment (RaymondHuang210129) [reply src/camera.ixx:66]: I would skip this issue as for now because the current camera visualization result does not have observable flaw.

**Comment by RaymondHuang210129** [comment]
Update: Smart pointers are applied. Revised node growing logic so that the parent's view is retained in one of the screen after splitted.

**Review by seanwu1105 — APPROVED** [review]
This PR is really good. Generalizing the sphere shader really improves the reusability.

I want to apologize for the problematic implementation of the quad tree. Originally, I want to make it fully immutable and pure functionally. However, the goal is totally defeated after I try to bind the `FirstPersonController` to each node as the controllers are fully stateful. This of course leads to many tricky bugs. Many thanks for fixing them. 

I always try to avoid from using pointers (even smart pointers) if the performance is not a problem. I still believe we can still achieve the same goal in this case without a pointer but that would require a lot of refactoring and redesign, which is meaningless for now. I believe the current way to patch the issue with new camera features is the best and neatest. What a great job.

**Review by JoannaWang1216 — APPROVED** [review]

**Review by seanwu1105 — APPROVED** [review]

**Review by sdCaramel302812 — APPROVED** [review]

**Pull request merged**